src/bidiMapper/modules/session/SubscriptionManager.ts

```diff
@@ -232,164 +189,204 @@ export class SubscriptionManager {
+   *
+   * This is a legacy spec branch to unsubscribe by attributes.
    */
-  unsubscribeAll(
-    events: ChromiumBidi.EventNames[],
-    contextIds: (BrowsingContext.BrowsingContext | null)[],
+  unsubscribe(
+    inputEventNames: ChromiumBidi.EventNames[],
+    inputContextIds: BrowsingContext.BrowsingContext[],
     channel: BidiPlusChannel,
   ) {
-    // Assert all contexts are known.
-    for (const contextId of contextIds) {
-      if (contextId !== null) {
-        this.#browsingContextStorage.getContext(contextId);
+    const eventNames = new Set(unrollEvents(inputEventNames));
+
+    for (const contextId of inputContextIds) {
+      // Validation that contexts exist.
+      this.#browsingContextStorage.getContext(contextId);
+    }
+    const topLevelTraversables = new Set(
+      inputContextIds.map((contextId) => {
+        const topLevelContext =
+          this.#browsingContextStorage.findTopLevelContextId(contextId);
+        if (!topLevelContext) {
+          throw new NoSuchFrameException(
+            `Top-level navigable not found for context id ${contextId}`,
+          );
+        }
+        return topLevelContext;
+      }),
+    );
+
+    const newSubscriptions: Subscription[] = [];
+    const isGlobal = topLevelTraversables.size === 0;
+    let found = false;
+    const eventsMatched = new Set<ChromiumBidi.EventNames>();
+    for (const subscription of this.#subscriptions) {
+      if (found || subscription.channel !== channel) {
+        newSubscriptions.push(subscription);
+        continue;
+      }
+      if (intersection(subscription.eventNames, eventNames).size === 0) {
+        newSubscriptions.push(subscription);
+        continue;
+      }
+      if (isGlobal) {
+        // If it is non-global subscription, keep it as is because we
+        // are only unsubscribing globally.
+        if (subscription.topLevelTraversableIds.size !== 0) {
+          newSubscriptions.push(subscription);
+          continue;
+        }
+        const subscriptionEventNames = new Set(subscription.eventNames);
+        for (const eventName of eventNames) {
+          if (subscriptionEventNames.has(eventName)) {
+            eventsMatched.add(eventName);
+            subscriptionEventNames.delete(eventName);
+          }
+        }
+        // If some events remain in the subscription, we keep it.
+        if (subscriptionEventNames.size !== 0) {
+          subscription.eventNames = subscriptionEventNames;
+          newSubscriptions.push(subscription);
+        }
+      } else {
+        // Skip global subscriptions.
+        if (subscription.topLevelTraversableIds.size === 0) {
+          newSubscriptions.push(subscription);
+          continue;
+        }
+        // If event name sets are not exactly the same,
+        // it is not a match for unsubscribe.
+        if (!equal(eventNames, subscription.eventNames)) {
+          newSubscriptions.push(subscription);
+          continue;
+        }
```

Comment:
I'm not sure if this is correct. Does it mean that after subscribe(a, b), calling unsubscribe(a) does not have any effect? I don't think it matches to the specification.

Reply:
Updated.

Reply:
I wonder why unit test didn't catch this case